### PR TITLE
[Core] support scan metrics

### LINKED ISSUE #2000: [Feature] Support scan metrics
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/incubator-paimon/issues) and found nothing similar.

### Motivation

Subtask of [PIP-3](https://cwiki.apache.org/confluence/display/PAIMON/PIP-3%3A+Introduce+metrics+for+Paimon)

### Solution

_No response_

### Anything else?

_No response_

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by schnappi17**
### Purpose

Linked issue: close #2000

### CODE REVIEW

**Comment by schnappi17** [comment]
Close this by refactoring the metric system: https://github.com/apache/incubator-paimon/pull/2169

File: paimon-core/src/main/java/org/apache/paimon/operation/AbstractFileStoreScan.java [hunk 1]
```diff
@@ -237,19 +249,26 @@ private Pair<Snapshot, List<ManifestEntry>> doPlan(
...
+
+        AtomicLong cntEntries = new AtomicLong(0);
         Iterable<ManifestEntry> entries =
                 ParallellyExecuteUtils.parallelismBatchIterable(
                         files ->
                                 files.parallelStream()
                                         .filter(this::filterManifestFileMeta)
                                         .flatMap(m -> readManifest.apply(m).stream())
                                         .filter(this::filterByStats)
+                                        .peek(e -> cntEntries.getAndIncrement())
```
Comment (tsreaper) [inline paimon-core/src/main/java/org/apache/paimon/operation/AbstractFileStoreScan.java:263]: Why not directly increase `cntEntries` by the size of the collected list? If you use `peek`, `cntEntries` will be increased by every element in this list. Note that this stream is parallelized, so there will be many concurrent writes which affect the performance.

File: paimon-core/src/main/java/org/apache/paimon/table/system/AuditLogTable.java [hunk 2]
```diff
@@ -247,6 +248,12 @@ public SnapshotReader withBucketFilter(Filter<Integer> bucketFilter) {
             return this;
         }
 
+        @Override
+        public SnapshotReader withMetricRegistry(MetricRegistry registry) {
+            // won't register metric
```
Comment (tsreaper) [inline paimon-core/src/main/java/org/apache/paimon/table/system/AuditLogTable.java:253]: Why not? `AuditLogTable` also scans from data file.

**Review by tsreaper — COMMENTED** [review]
Can you also bridge this metrics with Flink, just like `CommitMetrics`?

**Pull request merged**